src/full/Agda/Compiler/ToTreeless.hs

```diff
@@ -326,7 +326,15 @@ withContextSize n cont = do
   diff <- asks (((n -) . length) . ccCxt)
   if diff >= 1 then createLambdas diff cont else do
     let diff' = -diff
-    cxt <- shift diff . drop diff' <$> asks ccCxt
+    cxt <- -- shift diff .
+       -- Andreas, 2021-04-10, issue #5288
+       -- The @shift diff@ is wrong, since we are returning to the original
+       -- context from @cont@, and then we would have to reverse
+       -- the effect of @shift diff@.
+       -- We need to make sure that the result of @cont@ make sense
+       -- in the **present** context, not the changed context
+       -- where it is constructed.
+       drop diff' <$> asks ccCxt
```

Comment:
This looks correct, but I can't quite follow the explanation. Here's my understanding:

We are building a `TTerm` case tree from `CompiledClauses`. In order to be able to match we bind all variables we'll need in a top-level lambda `λ a b c d → ..` (say). As we compute the `TTerm` we keep a context (list) of `TTerm` deBruijn indices for each `CompiledClause` variable. This is a renaming from the *source* context of the `CompiledClause` to the *target* context of the `TTerm`.

After some pattern matching we might have
```
λ a b c d →
  case c of
    e :: f → {cxt = [d, f, e, b, a]}
```
Now, what's causing the problems here is that `CompiledClauses` can be underapplied, so you might have matched on a variable only to find that in the catch-all the variable you matched on is bound in a lambda in the right-hand side! Extending the example, we might have `CompiledClauses` looking like this:
```
case 2 of
  _::_ → done[d, f, e, b, a] ...
  _    → done[b, a] (λ c d → ...)
```
When we get to the catch-all, the context will be `[d, c, b, a]` but the right-hand side is only expecting `a` and `b` to be bound. What we need to do is compile the right-hand side and then apply it to the variables `c` and `d` that we already bound. This is what `withContextSize` does.

Crucially (and this is where the bug was), we are not changing the target context, only the source context (we want a `TTerm` that makes sense at this point). This means that the correct move is to drop the entries for the additional source variables, but not change what target variables the remaining source variables map to. Hence, `drop` but no `shift`.

Reply:
Thanks, I added your explanation under mine, leaving it to Hegel's return to synthesize thesis and anti-thesis.  (Joking... my explanation was very elliptical, yours is very detailed.)